File: backend/utils/security_utils.py
import secrets
import random, string
import jwt
from flask import request
from config import Config
from functools import wraps



def generate_random_state(length=16):
    """
    Generates a random `state` string for OAuth (CSRF protection).
    """
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))


def token_required(f):
    """
    Decorator requiring a valid token from the session_token cookie.
    """
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.cookies.get("session_token")
        if not token:
            return {"error": "No Authorization token provided"}, 401
        try:
            decoded = jwt.decode(token, Config.SECRET_KEY, algorithms=["HS256"])
            request.user_id = decoded.get("user_id")
            if not request.user_id:
                raise Exception("No user_id found in the token.")
        except jwt.ExpiredSignatureError:
            return {"error": "Token has expired"}, 401
        except jwt.InvalidTokenError:
            return {"error": "Invalid token"}, 401
        except Exception:
            return {"error": "Unauthorized access. Please log in again."}, 401

        return f(*args, **kwargs)
    return decorated

File: backend-node/src/utils/securityUtils.js
const jwt = require("jsonwebtoken");
const config = require("../config/env");

/**
 * Generates a random OAuth state string.
 */
const generateRandomState = (length = 16) => {
  const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  return Array.from({ length }, () => characters[Math.floor(Math.random() * characters.length)]).join("");
};

/**
 * Middleware to verify JWT token from cookies.
 */
const tokenRequired = (req, res, next) => {
  const token = req.cookies.session_token;
  if (!token) return res.status(401).json({ error: "No authorization token provided" });

  try {
    const decoded = jwt.verify(token, config.JWT_SECRET_KEY);
    req.user_id = decoded.user_id;
    if (!req.user_id) throw new Error("No user_id found in token.");
    next();
  } catch (error) {
    res.status(401).json({ error: "Invalid or expired token." });
  }
};

module.exports = { generateRandomState, tokenRequired };
